pkg/generators/models/testdata/cases/validation/generated/model_person.go

```diff
@@ -0,0 +1,218 @@
+// This file is auto-generated, DO NOT EDIT.
+//
+// Source:
+//     Title: Test
+//     Version: 0.1.0
+package generatortest
+
+import (
+	validation "github.com/go-ozzo/ozzo-validation/v4"
+	"github.com/go-ozzo/ozzo-validation/v4/is"
+
+	"time"
+)
+
+// Person is an object.
+type Person struct {
+	// Age:
+	Age float32 `json:"age,omitempty"`
+	// Base64:
+	Base64 string `json:"base64,omitempty"`
+	// Date:
+	Date string `json:"date,omitempty"`
+	// Datetime:
+	Datetime time.Time `json:"datetime,omitempty"`
+	// Email:
+	Email string `json:"email,omitempty"`
+	// Gender:
+	Gender Gender `json:"gender"`
+	// Hostname:
+	Hostname string `json:"hostname,omitempty"`
+	// Ip:
+	Ip string `json:"ip,omitempty"`
+	// Ipv4:
+	Ipv4 string `json:"ipv4,omitempty"`
+	// Ipv6:
+	Ipv6 string `json:"ipv6,omitempty"`
+	// Name:
+	Name string `json:"name"`
+	// Url:
+	Url string `json:"url,omitempty"`
+	// Uuid:
+	Uuid string `json:"uuid,omitempty"`
+}
+
+// Validate implements basic validation for this model
+func (m Person) Validate() error {
+	return validation.Errors{
+		"age": validation.Validate(
+			m.Age, validation.Min(float32(18)), validation.Max(float32(120)),
+		),
+		"base64": validation.Validate(
+			m.Base64, is.Base64,
+		),
+		"date": validation.Validate(
+			m.Date, validation.Date("2006-01-02"),
+		),
+		"datetime": validation.Validate(
+			m.Datetime, validation.Date(time.RFC3339),
+		),
+		"email": validation.Validate(
+			m.Email, is.Email,
+		),
+		"gender": validation.Validate(
+			m.Gender, InKnownGender,
+		),
+		"hostname": validation.Validate(
+			m.Hostname, is.Host,
+		),
+		"ip": validation.Validate(
+			m.Ip, is.IP,
+		),
+		"ipv4": validation.Validate(
+			m.Ipv4, is.IPv4,
+		),
+		"ipv6": validation.Validate(
+			m.Ipv6, is.IPv6,
+		),
+		"name": validation.Validate(
+			m.Name, validation.Length(2, 32),
```

Comment:
I think the string and array types should include a `validation.Required` when a minimum is specified and none empty

for example, this test is failing
```go
	require.Error(t, validation.Validate(&Person{
		Age:      120,
		Gender:   GenderDefault,
		Name:     "",
		Hostname: "localhost",
		Email:    "foo@localhost.com",
		Uuid:     uuid.NewV4().String(),
	}))
```

```sh
--- FAIL: TestPerson (0.12s)
    model_person_test.go:21:
        	Error Trace:	model_person_test.go:21
        	Error:      	An error is expected but got nil.
        	Test:       	TestPerson
FAIL
FAIL	generatortest	0.391s
FAIL
```

Reply:
I'm about to add this behavior. What about when we have string fields that have a format (like emall)? If they are not optional (they are in the required list) should we allow empty strings?

Reply:
@LucasRoesler This seems to be the main question here. Even our enum rule accepts empty strings as valid. Even when its not strictly following the spec, I would propose that we will just require a non empty value for nearly all cases where we have any other validation in place. The only places where I would leave it out would be where we have an explicit inclusion of the empty value -> Min() < 0 or MinLen() == 0. This will just feel more right.